### PR TITLE
Add Bundler v1 Deprecation Warning

### PR DESCRIPTION
**Pull request by kbukum1**
### What are you trying to accomplish?

The goal of this PR is to enhance the file parser to return the `package_manager` information directly. This change has been implemented for Bundler, and other package managers can have this feature added in the future. This information will be used to check for deprecation warnings specifically for Bundler v1.

### What issues does this affect or fix?

This change allows us to better handle deprecation warnings for Bundler v1 by providing a consistent way to access package manager information directly from the file parser.

### Anything you want to highlight for special attention from reviewers?

- The abstract `package_manager` method has been added to the base file parser.
- For Bundler, the `package_manager` method has been implemented and returns the necessary information for deprecation warnings.
- For other package managers, the `package_manager` method is currently not implemented but can be extended in the future.
- Ensure to review the implementation of the abstract method and its integration with the Bundler package manager.

### How will you know you've accomplished your goal?

- The file parser will now return the `package_manager` information directly.
- The Bundler implementation correctly returns the package manager information.
- Deprecation warnings for Bundler v1 can be effectively checked using this new context.
- Tests should confirm that the new functionality works correctly and does not break existing functionality.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: common/lib/dependabot/package_manager.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+require "sorbet-runtime"
+
+module Dependabot
+  class PackageManagerBase
+    extend T::Sig
+    extend T::Helpers
+
+    abstract!
+
+    # The name of the package manager (e.g., "bundler")
```
Comment (abdulapopoola) [inline common/lib/dependabot/package_manager.rb:13]: I was wondering if as part of this effort we make these constants too?

File: common/lib/dependabot/package_manager.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    sig { abstract.returns(String) }
+    def name; end
+
+    # The version of the package manager (e.g., "2.1.4")
+    sig { abstract.returns(T.nilable(String)) }
```
Comment (abdulapopoola) [inline common/lib/dependabot/package_manager.rb:18]: For consistency too; should this be the `Version` object we already have in dependabot?

Comment (abdulapopoola) [reply common/lib/dependabot/package_manager.rb:18]: That way we can simplify and standardize - e.g. there wouldn't be a need for the major_version method

File: common/lib/dependabot/shared_helpers.rb [hunk 3]
```diff
@@ -484,6 +484,95 @@ def self.helper_subprocess_bash_command(command:, stdin_data:, env:)
       env_keys = env ? env.compact.map { |k, v| "#{k}=#{v}" }.join(" ") + " " : ""
       "$ cd #{Dir.pwd} && echo \"#{escaped_stdin_data}\" | #{env_keys}#{command}"
     end
+
```
Comment (jake-inverted) [inline common/lib/dependabot/shared_helpers.rb:487]: Let's move these new methods into the PackageManager class, the SharedHelpers are more for executing shell commands and doing things with the file system.

File: common/lib/dependabot/shared_helpers.rb [hunk 4]
```diff
@@ -484,6 +484,95 @@ def self.helper_subprocess_bash_command(command:, stdin_data:, env:)
...
+    sig do
```
Comment (jakecoffman) [inline common/lib/dependabot/shared_helpers.rb:488]: Let's move these under the PackageManager class, the SharedHelpers are more for executing shell commands and file system helpers.

File: common/lib/dependabot/pull_request_creator/message_builder.rb [hunk 5]
```diff
@@ -80,15 +83,17 @@ class MessageBuilder
           dependency_group: T.nilable(Dependabot::DependencyGroup),
           pr_message_max_length: T.nilable(Integer),
           pr_message_encoding: T.nilable(Encoding),
-          ignore_conditions: T::Array[T::Hash[String, String]]
+          ignore_conditions: T::Array[T::Hash[String, String]],
+          notices: T.nilable(T::Array[T::Hash[T.any(String, Symbol), T.untyped]])
```
Comment (jakecoffman) [inline common/lib/dependabot/pull_request_creator/message_builder.rb:87]: Let's make Notice a class so it can be typed. The T::Hash pattern is prone to typos.

File: updater/lib/dependabot/updater/operations/update_all_versions.rb [hunk 6]
```diff
@@ -250,24 +320,28 @@ def log_requirements_for_update(requirements_to_unlock, checker)
...
+        sig { params(dependency_name: String, updated_deps: T::Array[Dependabot::Dependency]).returns(T::Boolean) }
         def peer_dependency_should_update_instead?(dependency_name, updated_deps)
           updated_deps
             .reject { |dep| dep.name == dependency_name }
             .any? do |dep|
               next true if existing_pull_request([dep])
-              next false if dep.previous_requirements.nil?
+
+              dep_previous_requirements = dep.previous_requirements
+              next false if dep_previous_requirements.nil?
```
Comment (jakecoffman) [inline updater/lib/dependabot/updater/operations/update_all_versions.rb:331]: It seems like some unrelated changes like these and the types for all the methods have gotten into this PR? Try to keep extra changes out since it's already quite large of a change.

File: bundler/lib/dependabot/bundler/package_manager.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+      sig { override.returns(T::Boolean) }
+      def deprecated
+        deprecated_versions.include?(version)
+      end
+
+      sig { override.returns(T::Boolean) }
+      def unsupported
+        !deprecated && version < supported_versions.first
+      end
```
Comment (jakecoffman) [inline bundler/lib/dependabot/bundler/package_manager.rb:51]: Rubyism

```suggestion
      sig { override.returns(T::Boolean) }
      def deprecated?
        deprecated_versions.include?(version)
      end

      sig { override.returns(T::Boolean) }
      def unsupported?
        !deprecated && version < supported_versions.first
      end
```

File: common/lib/dependabot/notices.rb [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    attr_reader :details
+
+    sig do
+      params(
+        mode: String,
+        type: String,
+        package_manager_name: T.nilable(String),
+        message: T.nilable(String),
+        markdown: T.nilable(String),
+        details: T::Hash[Symbol, T.untyped]
```
Comment (jakecoffman) [inline common/lib/dependabot/notices.rb:26]: I'm trying to understand why we have a `details` Hash? Message and Markdown are already instance variables that hold the same data.

File: common/lib/dependabot/notices.rb [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+    sig { returns(T::Hash[Symbol, T.untyped]) }
+    attr_reader :details
+
+    sig do
+      params(
+        mode: String,
+        type: String,
+        package_manager_name: T.nilable(String),
+        message: T.nilable(String),
```
Comment (abdulapopoola) [inline common/lib/dependabot/notices.rb:24]: can we make this String instead of nilable String? I'm assuming we can then check for empty strings and eliminate the chance of nilerrors

File: common/lib/dependabot/package_manager.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    abstract!
+
+    # The name of the package manager (e.g., "bundler").
+    # @example
+    #   package_manager.name #=> "bundler"
+    sig { abstract.returns(String) }
+    def name; end
+
+    # The version of the package manager (e.g., "2.1.4").
+    # Returns nil if the version is not set.
```
Comment (abdulapopoola) [inline common/lib/dependabot/package_manager.rb:20]: Wondering if we can make this stricter? To always return Versions instead of nilable Versions

File: common/lib/dependabot/notices.rb [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+    # Generates a message for supported versions.
+    # @param supported_versions [Array<Dependabot::Version>, nil] The supported versions of the package manager.
+    # @param support_later_versions [Boolean] Whether later versions are supported.
+    # @return [String, nil] The generated message or nil if no supported versions are provided.
+    sig do
+      params(
+        supported_versions: T.nilable(T::Array[Dependabot::Version]),
+        support_later_versions: T::Boolean
+      ).returns(T.nilable(String))
```
Comment (abdulapopoola) [inline common/lib/dependabot/notices.rb:64]: not sure; is it possible to avoid nilable here too?

File: updater/lib/dependabot/updater/operations/update_all_versions.rb (hunk 6 above)
Comment (jakecoffman) [reply updater/lib/dependabot/updater/operations/update_all_versions.rb:332]: I'm not sure why we needed to fully type this file, it's better to keep a PR focussed on doing one thing. Often typing a file can cause unrelated problems. Partially typing new methods is preferred in this case.

File: common/lib/dependabot/notices.rb (hunk 8 above)
Comment (jakecoffman) [reply common/lib/dependabot/notices.rb:31]: Let's keep this change focused on adding a warning to the PR body in the Updater and discuss how we'll add the warning in the UI next.

Comment (kbukum1) [reply common/lib/dependabot/notices.rb]: removed details from the notice.

File: updater/lib/dependabot/updater/operations/update_all_versions.rb (hunk 6 above)
Comment (kbukum1) [reply updater/lib/dependabot/updater/operations/update_all_versions.rb]: I just replaced it to make it T.must since above we already have nil check.

**Review by jakecoffman — APPROVED** [review]
Nice work!

You can run a CLI job with a Bundler 1 repo and the experiment enabled as a sanity test, if you haven't already.

**Comment by kbukum1** [comment]
> Nice work!
> 
> You can run a CLI job with a Bundler 1 repo and the experiment enabled as a sanity test, if you haven't already.

Thanks this is be my first time for for enabling experience.
